content/news/2022-01-06-bevy-0.6/index.md

```diff
@@ -0,0 +1,1007 @@
+
+![clusters](clusters.png)
+
+Due to WebGL2's lack of storage buffers, in the interest of cross-platform compatibility, the current implementation is limited to at most 256 lights to ensure lights can fit within a uniform buffer. We can support many more lights on other platforms by using storage buffers, which we will add support for in a future release.
+
+[Click here](https://youtu.be/dElYzzNovEk) for a video that illustrates Bevy's clustered forward rendering.
+
+### Sprite Batching
+
+<div class="release-feature-authors">authors: @cart</div>
+
+Sprites are now rendered in batches according to their texture within a z-level. They are also opportunistically batched across z-levels. This yields significant performance wins because it drastically reduces the number of draw calls required. Combine that with the other performance improvements in the new Bevy Renderer and things start to get very interesting! On my machine, the old Bevy renderer generally started dropping below 60fps at around 8,000 sprites in our "bevymark" benchmark. With the new renderer on that same machine I can get about 120,000 sprites!
+
+![bevymark](bevymark.png)
+
+
+### Sprite Ergonomics
+
+Sprite entities are now simpler to spawn:
+
+```rust
+// Old (Bevy 0.5)
+fn spawn_sprite(
+    mut commands: Commands,
+    asset_server: Res<AssetServer>,
+    mut materials: ResMut<Assets<ColorMaterial>>,
+) {
+    let texture_handle = asset_server.load("player.png");
+    commands.spawn_bundle(SpriteBundle {
+        material: materials.add(texture_handle.into()),
+        ..Default::default()
+    });
+}
+
+// New (Bevy 0.6)
+fn spawn_sprite(mut commands: Commands, asset_server: Res<AssetServer>) {
+    commands.spawn_bundle(SpriteBundle {
+        texture: asset_server.load("player.png"),
+        ..Default::default()
+    });
+}
+```
+
+No need to manage sprite materials! Their texture handle is now a direct component and color can now be set directly on the {{rust_type(type="struct" crate="bevy_sprite" version="0.6.0" name="Sprite")}} component.
+
+### WGSL Shaders
+
+Bevy now uses [WGSL](https://www.w3.org/TR/WGSL/) for our built-in shaders and examples. WGSL is a new shader language being developed for WebGPU (although it is a "cross platform" shader language just like GLSL). Bevy still supports GLSL shaders, but WGSL is nice enough that, for now, we are treating it as our "officially recommended" shader language. WGSL is still being developed and polished, but given how much investment it is receiving I believe it is worth betting on. Consider this the start of the "official Bevy shader language" conversation, not the end of it.
+
+```rust
+[[group(0), binding(0)]]
+var<uniform> view: View;
+
+[[group(1), binding(0)]]
+var<uniform> mesh: Mesh;
+
+struct Vertex {
+    [[location(0)]] position: vec3<f32>;
+};
+
+struct VertexOutput {
+    [[builtin(position)]] clip_position: vec4<f32>;
+};
+
+[[stage(vertex)]]
+fn vertex(vertex: Vertex) -> VertexOutput {
```

Comment:
Hmm on the one hand ... I see you point. On the other hand: "vertex" and "fragment" are the common "non-main" conventions for those shader functions. The type of the input to the vertex function is a Vertex. The variable name for something of type `Vertex` should be `vertex`. WGSL requires the `[[stage(vertex)]]` annotation. Some of those things we don't have power over, and the ones we do I wouldn't want to change. Maybe we can ultimately find a way to omit the "stage(vertex)" annotation?

In the short term, maybe it would be better to do a fragment shader for the example.

Reply:
I don't want to push that hard on this because it is a small thing and not critical (it would honestly be fine if left alone), however

>"vertex" and "fragment" are the common "non-main" conventions for those shader functions.

It is worth noting that the W3C use `vert_main`, `frag_main`, `comp_main`, in their [entry point declaration docs](https://www.w3.org/TR/WGSL/#entry-point-decl), and sotrh uses `vs_main` & `fs_main` in their [learn-wgpu tutorial](https://sotrh.github.io/learn-wgpu/beginner/tutorial3-pipeline/#writing-the-shaders). - So the convention doesn't seem ubiquitous.

Reply:
Hmm you are right. I was thinking of (and emulating) godot's `void vertex()` and `void fragment()` [conventions](https://docs.godotengine.org/en/3.0/tutorials/shading/shading_language.html). I think I still prefer the clarity of `vertex` and `fragment`, but its good to know that theres a variety of options out there. I think I'll probably just leave this as-is for now.